File: algorithms/python3/120_triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        '''dynamic programming'''
        res=triangle[-1].copy()
        for i in range(len(triangle)-2,-1,-1):
            for j in range(len(triangle[i])):
                res[j]=min(res[j],res[j+1])+triangle[i][j]
        return res[0]

a=Solution()

nums=[
     [2],
    [3,4],
   [6,5,7],
  [4,1,8,3]
]

print(a.minimumTotal(nums))

File: algorithms/cpp/120_triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        if (triangle.size()==0) return 0;
        vector<int> temp= triangle.back();
        for (int i=triangle.size()-2;i>=0;i--) {
            for (int j=0;j<=i;j++) {
                temp[j]=min(triangle[i][j]+temp[j], triangle[i][j]+temp[j+1]);
            }
        }
        return temp[0];
        }
};
